clang/lib/StaticAnalyzer/Checkers/StdLibraryFunctionsChecker.cpp

```diff
@@ -2211,6 +2221,15 @@ void StdLibraryFunctionsChecker::initFunctionSummaries(
                   ErrnoNEZeroIrrelevant, GenericFailureMsg)
             .ArgConstraint(NotNull(ArgNo(0))));
 
+    // int pclose(FILE *stream);
+    addToFunctionSummaryMap(
+        "pclose", Signature(ArgTypes{FilePtrTy}, RetType{IntTy}),
+        Summary(NoEvalCall)
+            .Case({ReturnValueCondition(WithinRange, {{0, IntMax}})},
+                  ErrnoMustNotBeChecked, GenericSuccessMsg)
+            .Case(ReturnsMinusOne, ErrnoNEZeroIrrelevant, GenericFailureMsg)
+            .ArgConstraint(NotNull(ArgNo(0))));
+
```

Comment:
Quoting manpage: `on success, returns the exit status of the command`, consequently, this might be negative on success.

I believe the manpage for `exit(3)` should tell us what values can we expect as exit codes.
> The exit() function causes normal process termination and the
       least significant byte of status (i.e., status & 0xFF) is
       returned to the parent (see [wait(2)](https://man7.org/linux/man-pages/man2/wait.2.html)).

So, I believe any value could be an exit code, that is representable by a signed char (if CHAR_BITS == 8, then [-128,127])

Also add tests for the range for the result of the `pclose`.

Reply:
I have tried if the return value can be negative on both linux and MacOS, it seems `pclose` always returns positive values on success, even if the child process called `exit(-3)`.

Here are my test cases,

```
// father.c, gcc father.c -o father
#include <stdio.h>
#include <stdlib.h>
int main() {
	FILE *fp = popen("/tmp/child.out", "w");
	if (fp) {
		int r = pclose(fp);
		printf("%d\n", r);
	}
	return 0;
}
```

```
// child.c, gcc child.c -o child
#include <stdio.h>
#include <stdlib.h>
int main() {
	exit(-3);
	return -3;
}
```

And actually the `child` returns `64768(253 << 8)` to the father.

So I think my current conditions are OK.

Reply:
The POSIX documentation does not tell if it is possible to get negative value from `fclose` other than -1. Because -1 is reserved for indication of error, it should not be a regular return value (if yes, there should be mentioned that `errno` is not changed at success, to make detection of error possible). If negative non-error return value is allowed, it can not be -1.
It looks like that negative return values from the process appear as positive values like in the example above. The returned value is not the value passed to `exit`, that can be extracted by `WEXITSTATUS` (see `waitpid`).

Reply:
Okay. Thanks for digging into this.
Consider this thread resolved.

Reply:
@balazske What is your opinion?

I think we can keep current form:
1. Negative but non -1 return values are not mentioned in the POSIX document.
2. Negative return values are not supported on real world linux&MacOS.

There may be negative but non -1 return values on other platforms, however we currently choose a conservative way as current form.

Reply:
We can keep the current form.